src/H5Tinit_float.c

```diff
@@ -306,14 +306,17 @@ H5T__fix_order(int n, int last, int *perm, H5T_order_t *order)
     if (last <= 0)
         HGOTO_ERROR(H5E_DATATYPE, H5E_CANTINIT, FAIL, "failed to detect byte order");
 
-    /* We have at least three points to consider */
-    if (perm[last] < perm[last - 1] && perm[last - 1] < perm[last - 2]) {
+    if (perm[last] < perm[last - 1] &&
+        /* Only check perm[last - 2] if we have more than 2 points to consider */
+        ((last < 2) || (perm[last - 1] < perm[last - 2]))) {
         /* Little endian */
         *order = H5T_ORDER_LE;
         for (int i = 0; i < n; i++)
             perm[i] = i;
     }
-    else if (perm[last] > perm[last - 1] && perm[last - 1] > perm[last - 2]) {
+    else if (perm[last] > perm[last - 1] &&
+             /* Only check perm[last - 2] if we have more than 2 points to consider */
+             ((last < 2) || (perm[last - 1] > perm[last - 2]))) {
```

Comment:
The changes here fix an issue where this function was assuming it could access `perm[last - 2]` even though that was invalid and would also cause the `_Float16` type to be detected as VAX endian order